src/eos-updater-apply.c

```diff
@@ -73,41 +77,506 @@ get_test_osname (void)
+        return NULL;
+
+      if (!file)
+        break;
+
+      if (!read_flatpak_refs_from_file_into_ptr_array (file,
+                                                       autoinstall_flatpaks,
+                                                       cancellable,
+                                                       error))
+        return NULL;
+    }
+
+  return g_steal_pointer (&autoinstall_flatpaks);
+}
+
+static gboolean
+flatpaks_to_deploy_from_commit (OstreeRepo   *repo,
+                                const gchar  *checksum,
+                                GPtrArray   **out_flatpaks_this_commit_wants,
+                                GCancellable *cancellable,
+                                GError      **error)
+{
+  g_autofree gchar *checkout_directory_path = NULL;
+  g_autoptr(GFile) checkout_directory = NULL;
+  OstreeRepoCheckoutAtOptions options = { 0, };
+  g_autoptr(GPtrArray) autoinstall_flatpaks = NULL;
+
+  g_return_val_if_fail (out_flatpaks_this_commit_wants != NULL, FALSE);
+
+  checkout_directory = get_temporary_directory_to_checkout_in (error);
+  checkout_directory_path = g_file_get_path (checkout_directory);
+
+  /* Now that we have a temporary directory, checkout the OSTree in it
+   * at the /usr/share/eos-application-tools path. If it fails, there's nothing to
+   * read, otherwise we can read in the list of flatpaks to be auto-installed
+   * for this commit. */
+  options.subpath = "usr/share/eos-application-tools/flatpak-autoinstall.d";
+
+  if (!ostree_repo_checkout_at (repo,
+                                &options,
+                                -1,
+                                checkout_directory_path,
+                                checksum,
+                                cancellable,
+                                error))
+    {
+      eos_updater_remove_recursive (checkout_directory, NULL);
+      return FALSE;
+    }
+
+  autoinstall_flatpaks = list_flatpak_refs_from_d_directory (checkout_directory,
+                                                             cancellable,
+                                                             error);
+  eos_updater_remove_recursive (checkout_directory, NULL);
+
+  if (!ostree_repo_checkout_gc (repo, cancellable, error))
+    return FALSE;
+
+  *out_flatpaks_this_commit_wants = g_steal_pointer (&autoinstall_flatpaks);
+  return TRUE;
+}
+
+static FlatpakInstallation *
+get_flatpak_installation (GCancellable *cancellable, GError **error)
+{
+  const gchar *override_path = g_getenv ("EOS_UPDATER_TEST_FLATPAK_INSTALLATION_DIR");
+
+  if (override_path)
+    {
+      g_autoptr(GFile) override = g_file_new_for_path (override_path);
+      return flatpak_installation_new_for_path (override, TRUE, cancellable, error);
+    }
+
+  return flatpak_installation_new_system (cancellable, error);
+}
+
+static guint
+flatpak_ref_hash (gconstpointer ref)
+{
+  return g_str_hash (flatpak_ref_get_name ((FlatpakRef *) ref));
```

Comment:
Technically you should not need the explicit cast, but if you did, it would be better to use `FLATPAK_REF()` instead.